### DISCUSSION TITLE
Does pycharm connect to the sage interpreter in the docker container or connect to the python interpreter in the container

### DISCUSSION THREAD

**Discussion by XiaoHuaSheng110** [Crypto-TII/claasp#52] [Q&A] [answered]
Does pycharm connect to the sage interpreter in the docker container or connect to the python interpreter in the container

**Comment by AnaCaceres** [comment, marked as answer]
Hello Xiao! 
It depends on how you configure the interpreter in PyCharm, it does not come predefined so you need to set it for the project. You would need to specify the interpreter to be the one from the docker container in order for PyCharm to use it. The one that the docker container has is the sage interpreter.
Hope this helps!

**Reply by XiaoHuaSheng110** [reply]
Hello Ana，
thanks for replying me，My main idea is to create a file in it by myself, write code with sage library and project library, and run it successfully. I am a docker container connected remotely, and I connect to the python interpreter in the container to run the code I wrote myself. It reports an error. The test file in the library can run with the pytest framework, but what I want is to run the file I wrote in the normal way, it does not run in the sage way, I set the sagemath interpreter path in the container, but pycharm shows that it is unknown The interpreter, is there a problem with the path of my sagemath interpreter, can you help me to see?
 And I have another question，I want to make a small qt project based on this library, can this library only run through the sage command line, but not through pycharm？
![image-20230719221812377](https://github.com/Crypto-TII/claasp/assets/115353515/cccb2732-48de-4ec0-8ec8-16578b8ccf0f)
![Z15A4~JT1LCLOL5I8S%B M3](https://github.com/Crypto-TII/claasp/assets/115353515/7e784dfb-ff87-4116-af65-af20c28d5511)
![{(8B{V%PIE@V AHKJ0QXLV8](https://github.com/Crypto-TII/claasp/assets/115353515/fac8f2cb-0ab4-4cb0-90f7-2ed4081c7ce4)
![YS`E6)GN7HZ` IWUET$CGK0](https://github.com/Crypto-TII/claasp/assets/115353515/4eb1caf5-0579-47c3-9032-cb9978b0b7bb)
                                                                                                                                                                           BestWishes,

**Reply by mariags803** [reply]
Hey Xiao! Here you have a step by step guide on how to set the correct interpreter from the docker container.

[Set Python interpreter of Docker in PyCharm.pdf](https://github.com/Crypto-TII/claasp/files/12096551/Set.Python.interpreter.of.Docker.in.PyCharm.pdf)

**Reply by XiaoHuaSheng110** [reply]
hello ana！
I have successfully connected to the sage interpreter in the docker container, and can successfully run all the test files in the library, but they are all run in pytest or sage command line. I want to create a python file to call claasp and run it in python3 mode, because I also need to call QT, is this possible, because I want to borrow this library to write a small QT project, how to do this?
For example, the 2.py written by myself below, the method of calling the library, will report an error if it is run in python3 mode, but it can be put into the sage command line or run in pytest mode, how to solve it?

![9NYTXL6G1` 9U% 7S T$KB](https://github.com/Crypto-TII/claasp/assets/115353515/14aef480-ab0e-4081-afd2-b0ba983181fd)

![}B{UP4X(~RY(C{4 6AMH%_H](https://github.com/Crypto-TII/claasp/assets/115353515/a3d9d858-c4e4-46ca-8a4d-556fe5eaa045)

**Reply by p-huynh** [reply]
Hi @XiaoHuaSheng110, 
 
I created a file similar to yours and ran it inside the docker container using `sage 2.py` and `python3 2.py`, both seemed just fine. Here is the file (note the inclusion of the line `from sage.all import *` first):

```
from sage.all import *
from sage.crypto.sboxes import PRESENT
from claasp.cipher_modules.models.milp.utils.generate_sbox_inequalities_for_trail_search import sbox_inequalities

def generate_inequalities():
    sbox_ineq_milp = sbox_inequalities(PRESENT, algorithm="milp")
    print(sbox_ineq_milp, '\n')
    sbox_ineq_greedy = sbox_inequalities(PRESENT, algorithm="greedy")
    print(sbox_ineq_greedy)


generate_inequalities()
```

Could you please provide more details so that we can perhaps reproduce your error?
